src/render/src/handle.rs

```diff
@@ -0,0 +1,85 @@
+use std::marker::PhantomData;
+use std::sync::{Arc, mpsc};
+use std::{cmp, hash};
+
+use core;
+use Backend;
+
+pub(crate) enum Garbage<B: Backend> {
+    ShaderLib(B::ShaderLib),
+    Buffer(B::Buffer),
+    Image(B::Image),
+    RenderTargetView(B::RenderTargetView),
+    DepthStencilView(B::DepthStencilView),
+    ConstantBufferView(B::ConstantBufferView),
+    ShaderResourceView(B::ShaderResourceView),
+    UnorderedAccessView(B::UnorderedAccessView),
+    Sampler(B::Sampler),
+}
+
+pub(crate) type GarbageSender<B> = mpsc::Sender<Garbage<B>>;
+pub(crate) type GarbageReceiver<B> = mpsc::Receiver<Garbage<B>>;
+pub(crate) fn garbage_channel<B: Backend>() -> (GarbageSender<B>, GarbageReceiver<B>) {
+    mpsc::channel()
+}
+
+macro_rules! resource_handle {
+    ($handle:ident, $guard:ident, $resource:ident) => {
+        #[derive(Clone, Debug)]
+        struct $guard<B: Backend>(Option<B::$resource>, GarbageSender<B>);
+
+        impl<B: Backend> $guard<B> {
+            fn resource(&self) -> &B::$resource {
+                self.0.as_ref().unwrap()
+            }
+        }
+
+        impl<B: Backend> cmp::PartialEq for $guard<B>
+            where B::$resource: cmp::PartialEq
+        {
+            fn eq(&self, other: &$guard<B>) -> bool {
+                self.resource().eq(&other.resource())
+            }
+        }
+
+        impl<B: Backend> cmp::Eq for $guard<B>
+            where B::$resource: cmp::Eq
+        {}
+        
+        impl<B: Backend> hash::Hash for $guard<B>
+            where B::$resource: hash::Hash
+        {
+            fn hash<H: hash::Hasher>(&self, state: &mut H) {
+                self.resource().hash(state)
+            }
+        }
+
+        impl<B: Backend> Drop for $guard<B> {
+            fn drop(&mut self) {
+                self.1.send(Garbage::$resource(self.0.take().unwrap()))
+                    .unwrap_or_else(|e| {
+                        error!("Could not drop {}: {}", stringify!($resource), e)
+                    });
+            }
+        }
+
+        #[derive(Clone, Debug)] // TODO: PartialEq, Eq, Hash
+        pub struct $handle<B: Backend>(Arc<$guard<B>>);
+
+        impl<B: Backend> $handle<B> {
+            pub fn resource(&self) -> &B::$resource {
+                &self.0.resource()
+            }
+        }
+    }
+}
+
+resource_handle! { ShaderLib, ShaderLibGuard, ShaderLib }
```

Comment:
The macro parameters should be different, i.e. `{Image, ImageInfo}`.
The 3rd parameter matches the 1st, so no need to provide it.
The 2nd parameter could be avoided if we have `mod guard { //all the guards with the same name as their resources`, so that `guard::Image` is the one being a part of `Image` handle. This may require specifying all the handle types in a single macro invocation, so that you can open and close the `guard` mod once.